crates/llm-base/src/util.rs

```diff
@@ -60,14 +62,15 @@ impl TokenUtf8Buffer {
         }
     }
 
-    /// Adapt a `&str` callback so that it can be used in a `&[u8]` context.
+    /// Adapt an [InferenceResponse] callback so that it can be used in a `&[u8]`
+    /// context.
     pub fn adapt_callback<'a, E: std::error::Error + 'static>(
-        mut callback: impl FnMut(&str) -> Result<(), E> + 'a,
-    ) -> impl FnMut(&[u8]) -> Result<(), E> + 'a {
+        mut callback: impl FnMut(InferenceResponse) -> Result<InferenceFeedback, E> + 'a,
+    ) -> impl FnMut(&[u8]) -> Result<crate::inference_session::InferenceFeedback, E> + 'a {
         let mut buffer = Self::new();
         move |token| match buffer.push(token) {
-            Some(tokens) => callback(&tokens),
-            None => Ok(()),
+            Some(tokens) => callback(InferenceResponse::PromptToken(tokens)),
```

Comment:
Hm, might be a little counterintuitive for it to always be a `PromptToken`. I'd suggest adding that to the name of the function (it seems like we only ever use it in a prompt context) or adding a parameter to switch between prompt and inference tokens.

Reply:
Okay, I just pulled it into `inference_session.rs` and gave it a more appropriate name :+1: